src/rockstor/system/tailscale.py

```diff
@@ -0,0 +1,204 @@
+import time
+from typing import List
+
+from storageadmin.exceptions import RockStorAPIException
+from system.constants import TAILSCALE
+from system.network import enable_ip_forwarding, disable_ip_forwarding
+
+from system.osi import run_command
+from system.services import get_tailscale_status
+
+# Config options supported in webUI
+# (used to filter out eventual duplicates in ts_custom_config)
+TS_UI_SETTINGS = [
+    "accept_routes",
+    "advertise_exit_node",
+    "advertise_routes",
+    "exit_node",
+    "exit_node_allow_lan_access",
+    "hostname",
+    "reset",
+    "ssh",
+]
+logger = logging.getLogger(__name__)
+
+
+def tailscale_up(config: dict = None, timeout: int = None):
+    """Start tailscale
+    Wrapper around `tailscale up`. It first gets args from config (if any)
+    and constructs the tailscale up command accordingly.
+    Adds the timeout flag is specified (used to generate authURL during the
+    LOGIN action from the webUI).
+    """
+    # Construct tailscale up command
+    cmd = [
+        TAILSCALE,
+        "up",
+    ]
+    if config is not None:
+        cmd = construct_tailscale_up_command(config)
+        # Enable IP forwarding if needed (subnet router or exit node enabled)
+        if any(key in config for key in ("advertise_exit_node", "advertise_routes")):
+            enable_ip_forwarding(name="tailscale", priority=99)
+    # Add short timeout to just trigger the creation of authURL
+    if timeout:
+        cmd.append(f"--timeout={timeout}s")
+    return run_command(cmd, log=True, throw=False)
+
+
+def tailscale_down(config: dict = None):
+    """Stop tailscale
+    Wrapper around `tailscale down`.
+    """
+    cmd = [
+        TAILSCALE,
+        "down",
+    ]
+    o, e, rc = run_command(cmd, log=True, throw=False)
+    if config is not None:
+        if any(key in config for key in ("advertise_exit_node", "advertise_routes")):
+            disable_ip_forwarding(name="tailscale")
+    return o, e, rc
+
+
+def extract_param(param: str) -> str:
+    """Get and format the name of a custom parameter
+    Isolates the name of the parameter from its cli format and
+    returns it after brief formatting.
+
+    There are 2 types of params:
+      - in the form --<key>=<value>
+      - in the form --<key>
+
+    These would return, for instance:
+      - "--accept-risk=all" -> "accept_risk"
+      - "--shields-up" -> "shields_up"
+    """
+    if re.search("=", param) is not None:
+        param = param.split("=")[0]
+    param = param.split("--")[1]
+
+    # Tailscale cli params use "-" whereas Rockstor uses "_" for storage
+    param = re.sub("-", "_", param)
+
+    return param
+
+
+def validate_ts_hostname(config: dict) -> dict:
+    """Ensure hostname is alphanumeric with hyphens only
+    No special character (including hyphens) is allowed as a custom hostname.
+
+    "hostname": "rock-dev_@#~!$%^&*()+123"
+    should return
+    "hostname": "rock-dev-123"
+    """
+    if "hostname" in config:
```

Comment:
quick question, can a hostname not have unicode characters in it?  If that's the case, wouldn't some "inclusive" regex be better than excluding what we currently know as "special" characters"?

Reply:
Oh that is one very good question here. It's true that this validation is rather ugly at the moment...
To answer your question on unicode characters: I've tested using Tailscale's UI, which allows you to rename a hostname. This is a good way to test what is valid or not. I've pasted the following into the box: `München`, and it autocorrected to a hyphen:
![image](https://github.com/rockstor/rockstor-core/assets/30297881/ae5c3a1b-25e6-45d0-aefd-ddaf3aea139d)

It thus seems we can't use anything besides alphanumeric and `-`. I'll change to using a regex filtering/correction as you recommended.

Reply:
@FroggyFlox @Hooverdan96 Re:

> I'll change to using a regex filtering/correction as you recommended.

Given this is a first draft, and a large one at that, I'm keen to get this out as-is. So this could be an enhancement by way of a follow-up. It likely won't be the last with such a significant addition.

Building now from what we have currently with a view to getting this out to the testing channel. We are getting along in this testing phase but not at the end yet so still time for some additions/refinements before we start heading for the next stable.